Major Changes to weather widget

cshariq opened the pull request:
I have done some major changes to the weather widget to enhance it. These are the following changes:

- Using weather.com api to avoid Unknown Information error

- Uses bing maps api to get users location for more accurate weather and location

- Adds a new option to adjust update frequency

- Can now display neighbourhood, city or address thanks to bing maps api

- You can now see 10 day forecast by swiping on the widget

- You can open the weather app on mac os 13 and over by tapping the weather widget

- New icons and options to switch between Illustrated, default, outlined and filled icons

- There's now an icon only mode


_When looking at the files changed section you can filter .png and .json files as they're all the new icons/assets and don't require any checking._

cshariq committed:
Major Changes to weather widget

I have done some major changes to the weather api to enhance it. These are the following changes:

- Using weather.com api to avoid Unknown Information error

- Uses bing maps api to get users location for more accurate weather and location

- Adds a new option to adjust update frequency

- Can now display neighbourhood, city or address thanks to bing maps api

- You can now see 10 day forecast by swiping on the widget

- You can open the weather app on mac os 13 and over by tapping the weather widget

- There's now an icon only mode

pigigaldi commented:
Thank you for you hard work!
I'll try to review this asap :) 

Just a quick tip for the future, you should avoid putting all changes into a single commit. Next time you could try split changes across multiple commits, with meaningful commit message (see other commits for message structure). This would make the review process much easier! :)

cshariq commented:
Ok thanks for the feedback, i've just started using GitHub, in fact this is my 2nd pull request. I'll try to keep your tips in mind for future commits.

cshariq commented on Weather/WeatherRepository.swift:
```diff
@@ -8,41 +8,34 @@
+        locationManager.startMonitoringSignificantLocationChanges()
         NotificationCenter.default.addObserver(self, selector: #selector(updateCurrentCondition), name: .didChangeWidgetPreferences, object: nil)
-        timer = Timer.scheduledTimer(withTimeInterval: 3600, repeats: true, block: { [weak self] _ in
+        if Preferences[.UpdateFrequency] == "Fifteen"{timers = 900} else if Preferences[.UpdateFrequency] == "Thirty"{timers = 1800} else{timers = 3600}
```
Modify update frequency based on user settings.

cshariq commented on Weather/WeatherWidget.swift:
```diff
@@ -9,66 +9,151 @@
+        set(image: NSImage(named: NSImage.touchBarSearchTemplateName))
         super.didLoad()
+        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
+            NotificationCenter.default.post(name: .didChangeWidgetPreferences, object: nil)
```
Update info when pock loads.

cshariq commented on Weather/WeatherWidget.swift:
```diff
@@ -9,66 +9,151 @@
+            NotificationCenter.default.post(name: .didChangeWidgetPreferences, object: nil)
+        }
     }
     override func didTapHandler() {
```
Double tap, triple tap and single tap detection. Opens weather app when taped once. In the future double and triple tap will show info like UV Index, Feels Like and more.

cshariq commented on Weather/WeatherWidget.swift:
```diff
@@ -9,66 +9,151 @@
+        return
+
+    }
+    override func didSwipeRightHandler() {
```
Goes back a day when viewing 10 day forecast.

cshariq commented on Weather/WeatherWidget.swift:
```diff
@@ -9,66 +9,151 @@
+            }
+        }
+    }
+    override func didSwipeLeftHandler() {
```
Shows 10 day forecast when swiped left.

cshariq commented on Weather/WeatherWidget.swift:
```diff
@@ -9,66 +9,151 @@
-        } else if let systemIcon = NSImage(named: data.weather.icon) {
-            view.set(image: systemIcon)
+        let locality = data.weather.name
+        if Preferences[.ShowIconOnly] == false || !(Day == 0){
```
Shows icon only based on user settings. Intentionally shows information if not viewing current forecast.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+        var description = " "
+        var temp = "0"
+        var temp_F = "0"
+        let url_for_maps = URL(string:  "https://dev.virtualearth.net/REST/v1/Locations/\(latitude),\(longitude)?includeNeighborhood=1&include=ciso2&o=json&key=Amn3nivKYow4ej0mliZVLFonh7W6ZBOlQz8FVrH0AJbK6_wViio1P8mrDODAgs6U")!
```
Get's city, neighbourhood, address and other info based on coordinates using bing maps api.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+        let url_for_maps = URL(string:  "https://dev.virtualearth.net/REST/v1/Locations/\(latitude),\(longitude)?includeNeighborhood=1&include=ciso2&o=json&key=Amn3nivKYow4ej0mliZVLFonh7W6ZBOlQz8FVrH0AJbK6_wViio1P8mrDODAgs6U")!
+        let url_for_weaather = URL(string: "https://api.weather.com/v1/geocode/\(latitude)/\(longitude)/aggregate.json?apiKey=e45ff1b7c7bda231216c7ab7c33509b8&products=conditionsshort,fcstdaily10short,fcsthourly24short,nowlinks")!
+        let MapRequest = URLRequest(url: url_for_maps, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 20)
+        let GetMaps = URLSession.shared.dataTask(with: MapRequest) { Mapdata, response, error in
```
Get's city, neighbourhood, address and other info based on coordinates using bing maps api.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
                 return
             }
+//           f
+            let decoder = JSONDecoder()
```
Decode response received from bing maps and extract the info.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+                var neighborhood = address.neighborhood
+                var postalCode = address.postalCode
+                var countryRegionIso2 = address.countryRegionIso2
+                if Day == 0{
```
Checks user settings and provides the right info to display.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+        }
+        
+        let request2 = URLRequest(url: url_for_weaather, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 20)
+        let GetWeather = URLSession.shared.dataTask( with: request2) { Wdata, response, error in
```
Get's the weather using weather.com api.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+            let weather = try? decoder.decode(WeatherResponse.self, from: Wdata)
+            let observation = weather?.conditionsshort?.observation
+            let FutureForcast = weather?.fcstdaily10short?.forecasts?[Day]
+            if !(Day == 0){
```
Checks what day needs to be displayed.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+                if Preferences[.show_description]{
+                    switch units {
+                    case "celsius":
+                            if description.count > 19{
```
Adjust title and description formatting based on the length of the info.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+//            } catch {
+//                print("Error: \(error.localizedDescription)")
+//            }
+            let decoder = JSONDecoder()
```
Decodes response received from weather.com api.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+            }else if Preferences[.IconStyle] == "Filled"{
+                Icon = "2-" + String(DisplayIcon!)
+            }else if Preferences[.IconStyle] == "Default"{
+                if DisplayIcon == 011 || DisplayIcon == 012 || DisplayIcon == 013 || DisplayIcon == 014 || DisplayIcon == 016 || DisplayIcon == 019 || DisplayIcon == 020 || DisplayIcon == 021 || DisplayIcon == 022 || DisplayIcon == 026 || DisplayIcon == 027 || DisplayIcon == 028 || DisplayIcon == 029 || DisplayIcon == 030 || DisplayIcon == 031 || DisplayIcon == 032 || DisplayIcon == 036 || Displa ...
```
Checks if icon is available in default icon pack, if not it displays outlined icon pack.

cshariq commented on Weather/WeatherService.swift:
```diff
@@ -37,58 +142,500 @@ struct WeatherData: Codable {
+                DisplayIcon = observation?.wx_icon!
             }
+            print(String(DisplayIcon!))
+            if Preferences[.IconStyle] == "Outlined"{
```
Checks user preferences and sets the icon accordingly.

cshariq commented on Weather/cities.csv:
Since the location is taken using the device's coordinates and bing maps api this is no longer needed.

cshariq committed:
Update WeatherWidget.swift

Fixes a glitch where the view isn't automatically reset to the current weather/conditions after 10 seconds when swiped more than once.

cshariq committed:
Improved Functionallity

- New icon and temp mode
- New only temp mode
- See more info about the selected day's weather by double tapping the widget
- Bug Fixes
